File: library/src/main/java/org/yzjt/sdk/base/BaseFragment.java
package org.yzjt.sdk.base;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import org.yzjt.sdk.nozzle.OnSingleClickListener;
import org.yzjt.sdk.util.ViewUtil;

public abstract class BaseFragment extends Fragment implements View.OnClickListener, OnSingleClickListener {
    protected String TAG = getClass().getSimpleName();
    private AppCompatActivity mActivity;
    protected View mContentView;
    protected Context mContext;
    private long[] antiDoubleClick = new long[2];
    private OnSingleClickListener mSingleClickListner;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mContext = getActivity();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mContentView = getContentView(inflater,container);
        initView();
        mSingleClickListner = this;
        return mContentView;
    }

    protected abstract View getContentView(LayoutInflater inflater, ViewGroup container);

    protected abstract void initView();

    protected <T extends View> T findViewById(int resId){
        if(mContentView == null){
            return null;
        }
        return ViewUtil.findViewById(resId,mContentView);
    }

    /**
     * click事件,通过对外只暴露了OnSingleClickListener的监听器,
     * 让子类的activity去实现,这样防止用户的连点
     *
     * @param view
     */
    @Override
    public void onClick(View view) {
        antiDoubleClick[0] = antiDoubleClick[1];
        antiDoubleClick[1] = System.currentTimeMillis();
        if (antiDoubleClick[1] - antiDoubleClick[0] < 300) {
            Toast.makeText(view.getContext(), "请勿连续点击", Toast.LENGTH_SHORT).show();
        } else {
            if (mSingleClickListner != null) {
                mSingleClickListner.onSingleClick(view);
            }
        }
    }

    @Override
    public void onSingleClick(View view) {

    }
}

File: library/src/main/java/org/yzjt/sdk/mvvm/base/BaseFragment.kt
package org.yzjt.sdk.mvvm.base

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.annotation.CallSuper
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import org.yzjt.sdk.mvvm.view.MvvMView
import org.yzjt.sdk.mvvm.viewmodel.BaseViewModel
import org.yzjt.sdk.nozzle.OnSingleClickListener

/**
 * @author lt
 * @param <T> The type of data model
 * @param <VM> The type of view model
 */
abstract class BaseFragment<T,VM : BaseViewModel<T>> : Fragment(), MvvMView<T, VM>, OnSingleClickListener, View.OnClickListener  {

    val TAG = javaClass.simpleName
    protected var mContext: Context? = null
    protected lateinit var mContentView: View
    private val antiDoubleClick = LongArray(2)
    private var mSingleClickListener: OnSingleClickListener? = null
    protected var mViewModel:VM? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        mContext = activity
    }

    /**
     * @return layout id
     *
     * @param savedInstanceState
     */
    override fun getContentViewId(savedInstanceState: Bundle?): Int {
        throw UnsupportedOperationException("Please override this method and return a effective layout id.")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        init()
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mContentView = layoutInflater.inflate(getContentViewId(savedInstanceState),container)
        mViewModel = getViewModel()
        initView()
        return mContentView
    }

    @CallSuper
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mSingleClickListener = this
        initViewObservable()
        initData()
    }

    fun <T : View> findViewById(resId:Int){
        if(mContentView == null){
            return
        }
        mContentView.findViewById<T>(resId)
    }

    /**
     * click事件,通过对外只暴露了OnSingleClickListener的监听器,
     * 让子类的activity去实现,这样防止用户的连点
     *
     * @param view
     */
    override fun onClick(view: View) {
        antiDoubleClick[0] = antiDoubleClick[1]
        antiDoubleClick[1] = System.currentTimeMillis()
        if (antiDoubleClick[1] - antiDoubleClick[0] < 300) {
//            Toast.makeText(view.context, "请勿连续点击", Toast.LENGTH_SHORT).show()
        } else {
            mSingleClickListener?.onSingleClick(view)
        }
    }

    override fun onSingleClick(view: View) {

    }
}